cmd/virt-apiservice-register/virt-apiservice-register.go

```diff
@@ -0,0 +1,737 @@
+
+	// Retrieve the CSR:
+	csr, err = csrRsc.Get(name, metav1.GetOptions{})
+	return
+}
+
+// createCSR creates a key pair and a certificate singing request. It sends the request to the server
+// and returns the CSR and the bytes of the private key, in PEM format.
+//
+func createCSR(client *kubernetes.Clientset, name string, cn string, addrs []string) (csr *certv1.CertificateSigningRequest, key []byte, err error) {
+	// Generate the CSR:
+	csrBytes, key, err := makeCSR(cn, addrs)
+	if err != nil {
+		return
+	}
+
+	// Get the reference to the CSR resource:
+	csrRsc := client.CertificatesV1beta1().CertificateSigningRequests()
+
+	// Send the CSR to the server:
+	csr, err = csrRsc.Create(
+		&certv1.CertificateSigningRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: certv1.CertificateSigningRequestSpec{
+				Groups: []string{
+					"system:authenticated",
+				},
+				Request: csrBytes,
+				Usages: []certv1.KeyUsage{
+					certv1.UsageDigitalSignature,
+					certv1.UsageKeyEncipherment,
+					certv1.UsageServerAuth,
+				},
+			},
+		},
+	)
+
+	return
+}
+
+// findCN calculates the common name for the certificate. It returns the calculated CN as well as
+// other names that should also be included in the certificate.
+//
+func findCN() (cn string, names []string) {
+	// Currently the API aggregator creates HTTPS connections to the API services using incomplete
+	// DNS names in the URL. If the name of the service is 'myservice' it uses 'myservice.default.svc'
+	// when the correct DNS name should be 'myservice.default.svc.cluster.local'. This is probably a
+	// bug in the API aggreator. To work it around we will add to the set of names that will be
+	// included in the certificate both the correct and incorrect names.
+	names = []string{
+		fmt.Sprintf("%s.%s.svc.%s", serviceName, namespaceName, dnsDomain),
+		fmt.Sprintf("%s.%s.svc", serviceName, namespaceName),
+	}
+
+	// Select the first name as the CN:
+	cn = names[0]
+
+	return
+}
+
+// waitCSR waits till the CSR for the given common name has been approved, denied or removed. The
+// complete CSR will be returned unless there is an error.
+//
+func waitCSR(client *kubernetes.Clientset, name string) (csr *certv1.CertificateSigningRequest, err error) {
+	var ok bool
+
+	// Get the reference to the CSR resource:
+	csrRsc := client.CertificatesV1beta1().CertificateSigningRequests()
+
+	// Retrieve the latest state of the CSR and check if it is already approved or denied:
+	csr, err = csrRsc.Get(name, metav1.GetOptions{})
+	if err != nil {
+		return
+	}
+	if isCSRApproved(csr) || isCSRDenied(csr) {
+		return
+	}
+
+	// Watch for changes to the CSR, till it is approved, denied or removed:
+	csrWatch, err := csrRsc.Watch(
+		metav1.ListOptions{
+			FieldSelector:   "metadata.name=" + name,
+			ResourceVersion: csr.ResourceVersion,
+		},
+	)
+	if err != nil {
+		return
+	}
+	csrCh := csrWatch.ResultChan()
```

Comment:
what would you think about a timeout value for this wait? I'm not sure how I feel about waiting indefinitely here.